### ISSUE TITLE
Make 🤗 Diffusers tests device-agnostic

### ISSUE THREAD

**Issue by arsalanu** [huggingface/diffusers#5562]
## Feature Request

Make the Diffusers test suite device-agnostic, so alternative accelerator devices are able to easily run the repo tests.

### Motivation
At the moment, Diffusers test suite explicitly supports `cuda` and `mps` devices. #4673 has made this less explicit, and other devices supporting PyTorch can be enabled. However, further changes to the tests need to be made to fully enable testing with alternative accelerators. For example, CUDA specific functions are still used in certain tests (e.g., `torch.cuda.manual_seed`), and decorators such as `require_torch_gpu` are used for FP16 tests, where other accelerators may be able to run them. This support would make it much easier to test alternative accelerators with PyTorch backends with Diffusers.

We have made a similar change to the 🤗 Transformers library, see issue [#25654](https://github.com/huggingface/transformers/issues/25654) and PR [#25870](https://github.com/huggingface/transformers/pull/25870) and would like to implement a similar change, although more specific to the testing requirements of Diffusers. 

I have a pull request that is ready to go for Diffusers, but as it affects multiple files and potentially requires changes to almost every file in the test suite, I thought it was better to open this issue to seek any further feedback from maintainers as to whether this is a reasonable modification.

### Proposal
* Users can specify an alternative backend using `DIFFUSERS_TEST_BACKEND` to allow for any additional imports needed to make PyTorch use this backend (for example, `torch_xla` which requires an additional import alongside `torch` in order for torch to use XLA tensors.)
* Users can specify a device specification file by pointing to it with `DIFFUSERS_TEST_DEVICE_SPEC` - this file lets the user specify 'replacements' for device specific functions. The proposed specification file format is as follows (this is what is used for Transformers):
``` 
import torch
import torch_npu
# User can add additional imports here

# Specify the device name (eg. 'cuda', 'cpu')
DEVICE_NAME = 'npu'

# Specify device-specific backends to dispatch to.
# If not specified (i.e., `None`) will fallback to 'default' in 'testing_utils.py`
MANUAL_SEED_FN = torch.npu.manual_seed
EMPTY_CACHE_FN = None
DEVICE_COUNT_FN = torch.npu.device_count
SUPPORTS_TRAINING=True
```
* In `testing_utils.py`: Add device-agnostic functions which dispatch to a specified alternative backend. This is mainly applicable to device-specific functions/capabilities, e.g., `torch.cuda.manual_seed` being replaced with `backend_manual_seed(device, seed)`. The dispatching of these functions will be based on the definitions in the above spec file. 
* In `import_utils.py`: Add automatic checks for data type support, such as `is_torch_fp16_available()` these functions should check whether an accelerator supports the datatype and are used as checks in the decorator for the specific behaviour. At the moment, any FP16 test explicitly only runs on GPUs, stopping other backends from testing it.
* In `testing_utils.py`: Add device-agnostic test functions and decorators: `require_torch_accelerator()` should replace `require_torch_gpu()` in most cases, and some additional decorators such as `require_torch_accelerator_with_fp16` will be needed to handle behaviours which may vary over different devices (e.g. support for FP16, support for training).
* Functions and decorators are both useful, as decorators avoid running tests which focus on a specific behaviour, and functions directly replace functionality within tests to allow them to use all features of an accelerator.

In the tests themselves, decorators specific to `gpu` should only remain where the behaviour is truly GPU-specific (e.g., use of `xformers`). Otherwise, tests which can run on any accelerator should use `require_torch_accelerator()` as a decorator.

Lastly, it is important these changes do not affect any current CI runners in Diffusers, so behaviour for the supported devices (`cuda`, `cpu`, `mps`) will not be affected by these changes. The alternative function dispatch keeps defaults for these backends. These extra environment variables don't need to be specified if not using a custom accelerator.

**Short summary**: **Replace parts of code that are overly specific to one accelerator (e.g. CUDA functions) with a more generic dispatch function which can redirect to different backends. Add checks for behaviour which may vary across accelerators.**

### Considerations
We found that a Python file as a device specification was the most customisable way for the user to control device-agnostic functionality, as it allows easy specification of functions as variables, and allows the user to write in extra imports if needed. We considered just using more environment variables or alternatives like YAML to create the specification, but this seemed easiest for UX, and avoids having to parse strings, `getattr` etc., and mitigates syntax errors.

### Additional context
These changes are very similar to the changes made for Transformers. A similar implementation with device-agnostic dispatch for device-specific functions using a Python file and custom backend specification with environment variables is already functional and merged into Transformers.

**Comment by patrickvonplaten** [comment]
Thanks for opening the issue! Let me sync real quick with some folks

**Comment by patrickvonplaten** [comment]
Ok, I think we're in line with your idea :-) Thank for opening an issue first! 

It'd be great if you could open a PR, but to begin with could you maybe just apply the changes to a single test file so that we can discuss this and if it works we'll expand it to all test files?

**Comment by arsalanu** [comment]
Brilliant, thank you! 

Happy to make it affect only a single test for now, I'll get onto it shortly :+1:

**Comment by arsalanu** [comment]
Initial implementation is ready for review: #5612
